plugins/org.python.pydev.shared_ui/src/org/python/pydev/shared_ui/field_editors/JsonFieldEditor.java

```diff
@@ -0,0 +1,609 @@
+            GC gc = new GC(textField);
+            try {
+                if (widthInChars != UNLIMITED) {
+                    Point extent = gc.textExtent("X");//$NON-NLS-1$
+                    gd.widthHint = widthInChars * extent.x;
+                } else {
+                    gd.horizontalAlignment = GridData.FILL;
+                    gd.grabExcessHorizontalSpace = true;
+                }
+                if (heigthInChars > 1) {
+                    gd.heightHint = heigthInChars * gc.getFontMetrics().getHeight();
+                }
+            } finally {
+                gc.dispose();
+            }
+        } else {
+            gd.horizontalAlignment = GridData.FILL;
+            gd.grabExcessHorizontalSpace = true;
+        }
+        textField.setLayoutData(gd);
+    }
+
+    @Override
+    protected void doLoad() {
+        if (textField != null) {
+            String value = getPreferenceStore().getString(getPreferenceName());
+            textField.setText(value);
+            oldValue = value;
+        }
+    }
+
+    @Override
+    protected void doLoadDefault() {
+        if (textField != null) {
+            String value = getPreferenceStore().getDefaultString(
+                    getPreferenceName());
+            textField.setText(value);
+        }
+        valueChanged();
+    }
+
+    @Override
+    protected void doStore() {
+        getPreferenceStore().setValue(getPreferenceName(), textField.getText());
+    }
+
+    /**
+     * Returns the error message that will be displayed when and if
+     * an error occurs.
+     *
+     * @return the error message, or <code>null</code> if none
+     */
+    public String getErrorMessage() {
+        return errorMessage;
+    }
+
+    @Override
+    public int getNumberOfControls() {
+        return 2;
+    }
+
+    /**
+     * Returns the field editor's value.
+     *
+     * @return the current value
+     */
+    public String getStringValue() {
+        if (textField != null) {
+            return textField.getText();
+        }
+
+        return getPreferenceStore().getString(getPreferenceName());
+    }
+
+    /**
+     * Returns this field editor's text control.
+     *
+     * @return the text control, or <code>null</code> if no
+     * text field is created yet
+     */
+    protected StyledText getTextControl() {
+        return textField;
+    }
+
+    /**
+     * Returns this field editor's text control.
+     * <p>
+     * The control is created if it does not yet exist
+     * </p>
+     *
+     * @param parent the parent
+     * @return the text control
+     */
+    public StyledText getTextControl(Composite parent) {
+        if (textField == null) {
+            textField = createTextWidget(parent);
+            textField.setFont(parent.getFont());
+            switch (validateStrategy) {
+                case VALIDATE_ON_KEY_STROKE:
+                    textField.addKeyListener(new KeyAdapter() {
```

Comment:
I think that validation should be done in a thread (you can create a subclass of `org.eclipse.core.runtime.jobs.Job` and `schedule` it to run -- so, the current call that does the check will actually set the contents to validate in the job and `schedule` it... some care needs to be taken so that the contents are validated prior to accepting what was actually entered) -- I've seen some extreme examples of path mapping setup where there were thousands of lines here, in which case doing it at each keystroke would not work well.

Note that the job (which is actually a separate thread) should receive the text and do the validation, but it cannot access UI elements (as it'll not run in the UI thread), so, when finished it needs to do something as:

```
            // calculate everything in the thread, and when finished
            // run in the ui thread to set/clear error messages
            RunInUiThread.async(new Runnable() {

                public void run() {
                    if (disposed) {
                        return;
                    }
                    setErrorMessage(....);
                }
            });
```

Reply:
I don't know if I get it. By "schedule to run" you mean to set a keypress idle time handler? i.e.: it will only call the validation (in another thread) when the user typed something and then stopped key pressing for a little while. This is the idea?